### PR TITLE
Compute services and related functions.

### PR DESCRIPTION
**Pull request by emapco**
- Adds step functions and related functions for computing the next CA state. A step function was provided so that the user (user II) can provided their own custom rule function for computing a cell's new state.
- Adds guard clauses to handle failure modes. 
- Adds a method for printing a error message for a given error code.
- Adds a utility header and implementation file for auxiliary functionality that the `CellularAutomata` class uses.
- Updates the docstrings for better website formatting includes an updated documentation site.

### CODE REVIEW

**Comment by trevoroldham** [comment]
I've looked through the updates and it seems like good work, I will wait to let you merge this code after Chongye has a look at it. I got your message about the gravitational rule but I will have to discuss that more with you during our meeting this weekend. Thanks for all the work you did here.

File: Include/CAdatatypes.h [hunk 1]
```diff
@@ -84,12 +118,35 @@ class CellularAutomata // The CA data structure
...
+
+    // function for setting the cell state with generated neighborhood array; tensor case
+    int get_state_from_neighborhood_3d(int *cell_index, int index_size, int &new_cell_state,
+                                       void(custom_rule)(int *, int, int *, int, int &));
+
+    // simulates a CA step
+    int step();
+
+    // simulates a CA step with a custom_rule
+    int step(void(custom_rule)(int *, int, int *, int, int &));
```
Comment (TsukiTiger) [inline Include/CAdatatypes.h:145]: Cool!!

File: Include/CAdatatypes.h [hunk 2]
```diff
@@ -12,70 +12,104 @@
...
+    int ***next_tensor; //!< tensor for three dimensional grid of cells holding the next state
+    int steps_taken;    //!< the number of steps the CA has taken
+
 public:
+    /**
+     * @brief enum containing the different type of neighborhoods
+     * the CellularAutomata class supports
+     *
+     */
+    enum neighborhood
```
Comment (TsukiTiger) [inline Include/CAdatatypes.h:41]: I learned a lot of new contents!

**Review by TsukiTiger — COMMENTED** [review]
I reviewed the new codes you added. They are pretty neat and amazing. It is a good material for me to learn new syntax as well! Good Job!! Thank you!!

**Pull request merged**